File: graph/dfs/python/dfs_programiz.py
def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)  # 0

    for next in graph[start] - visited:  # {1, 2}
        dfs(graph, next, visited)  # {next: 1}
    return visited


graph = {'0': set(['1', '2']),
         '1': set(['0', '3']),
         '2': set(['0', '4', '5']),
         '3': set(['1']),
         '4': set(['2', '6']),
         '5': set(['2']),
         '6': set(['4'])}

#[[1, 2], [0, 3], [0, 4, 5], [1], [2, 6], [2], [4]]
v = dfs(graph, '0')
print(v)

{'0', '5', '2', '3', '6', '1', '4'}
[0, 2, 5, 4, 6, 1, 3]

File: graph/dfs/c++/dfs_programiz.cpp
#include <iostream>
#include <list>

using namespace std;

class Graph
{
    int numVertices;
    list<int> *adjLists;
    bool *visited;

public:
    Graph(int V);
    void addEdge(int src, int toEdge);
    void DFS(int root);
};

// initialize Graph

Graph::Graph(int V)
{
    numVertices = V;
    adjLists = new list<int>[V];
    visited = new bool[V];
}

void Graph::addEdge(int src, int toEdge)
{
    adjLists[src].push_front(toEdge);
}

void Graph::DFS(int vertex)
{
    visited[vertex] = true;
    list<int> adjList = adjLists[vertex];

    cout << vertex << " ";

    list<int>::iterator i;
    for (i = adjList.begin(); i != adjList.end(); ++i)
    {
        if (!visited[*i])
            DFS(*i);
    }
}

int main()
{
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 3);

    g.DFS(0);

    return 0;
}